models/imap.go

```diff
@@ -0,0 +1,151 @@
+package models
+
+import (
+	"errors"
+	"net"
+	"time"
+
+	log "github.com/gophish/gophish/logger"
+)
+
+const DefaultIMAPFolder = "INBOX"
+const DefaultIMAPFreq = 60 // Every 60 seconds
+
+// IMAP contains the attributes needed to handle logging into an IMAP server to check
+// for reported emails
+type IMAP struct {
+	UserId                      int64     `json:"-" gorm:"column:user_id"`
+	Enabled                     bool      `json:"enabled"`
+	Host                        string    `json:"host"`
+	Port                        uint16    `json:"port,string,omitempty"`
+	Username                    string    `json:"username"`
+	Password                    string    `json:"password"`
+	TLS                         bool      `json:"tls"`
+	Folder                      string    `json:"folder"`
+	RestrictDomain              string    `json:"restrict_domain"`
+	DeleteReportedCampaignEmail bool      `json:"delete_reported_campaign_email"`
+	LastLogin                   time.Time `json:"last_login,omitempty"`
+	ModifiedDate                time.Time `json:"modified_date"`
+	IMAPFreq                    uint32    `json:"imap_freq,string,omitempty"`
+}
+
+// ErrIMAPHostNotSpecified is thrown when there is no Host specified
+// in the IMAP configuration
+var ErrIMAPHostNotSpecified = errors.New("No IMAP Host specified")
+
+// ErrIMAPPortNotSpecified is thrown when there is no Port specified
+// in the IMAP configuration
+var ErrIMAPPortNotSpecified = errors.New("No IMAP Port specified")
+
+// ErrInvalidIMAPHost indicates that the IMAP server string is invalid
+var ErrInvalidIMAPHost = errors.New("Invalid IMAP server address")
+
+// ErrInvalidIMAPPort indicates that the IMAP Port is invalid
+var ErrInvalidIMAPPort = errors.New("Invalid IMAP Port")
+
+// ErrIMAPUsernameNotSpecified is thrown when there is no Username specified
+// in the IMAP configuration
+var ErrIMAPUsernameNotSpecified = errors.New("No Username specified")
+
+// ErrIMAPPasswordNotSpecified is thrown when there is no Password specified
+// in the IMAP configuration
+var ErrIMAPPasswordNotSpecified = errors.New("No Password specified")
+
+var ErrInvalidIMAPFreq = errors.New("Invalid polling frequency.")
+
+// TableName specifies the database tablename for Gorm to use
+func (s IMAP) TableName() string {
```

Comment:
Throughout this file it looks like there's multiple instances where we call an `IMAP` instance `s`, which probably isn't the best name for it. Could we change these to `im` and, for plural as needed, `ims`? That would be a bit more descriptive when we need to modify the code later on.

Reply:
Good point, changed.